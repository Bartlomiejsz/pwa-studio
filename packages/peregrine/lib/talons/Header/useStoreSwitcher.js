import { useQuery } from '@apollo/client';
import { Util } from '../../index';
const { BrowserPersistence } = Util;

/**
 *
 * @param {*} props.query the store switcher data query
 */
export const useStoreSwitcher = props => {
    const { query } = props;
    const { data } = useQuery(query);
    const storage = new BrowserPersistence();

    const availableStores = [...data.availableStores].reduce(
        (storeViews, store) => {
            storeViews[store.code] = {
                storeName: store['store_name'],
                locale: store.locale
            };
            return storeViews;
        },
        {}
    );

    const handleSwitchStore = storeCode => {
        storage.setItem('store_view_code', storeCode);
        console.log(storeCode);
    };

    return {
        handleSwitchStore,
        availableStores
    };
};
