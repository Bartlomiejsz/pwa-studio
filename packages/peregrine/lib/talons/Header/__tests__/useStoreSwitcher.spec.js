import React from 'react';
import { useHistory } from 'react-router-dom';

import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';
import { Util } from '@magento/peregrine';
import { useStoreSwitcher } from '../useStoreSwitcher';

const setItem = jest.fn();
jest.spyOn(Util, 'BrowserPersistence').mockImplementation(
    function BrowserPersistence() {
        return { setItem };
    }
);

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(() => ({}))
}));

const history = {
    go: jest.fn()
};

useHistory.mockImplementation(() => history);

jest.mock('@apollo/client', () => {
    const useQuery = jest.fn().mockReturnValue({
        data: {
            storeConfig: {
                code: 'store2'
            },
            availableStores: [
                {
                    code: 'store1',
                    locale: 'locale1',
                    store_name: 'Store 1',
                    default_display_currency_code: 'USD'
                },
                {
                    code: 'store2',
                    locale: 'locale2',
                    store_name: 'Store 2',
                    default_display_currency_code: 'EUR'
                }
            ]
        },
        error: null,
        loading: false
    });
    return { useQuery };
});

jest.mock('@magento/peregrine/lib/hooks/useDropdown', () => ({
    useDropdown: jest.fn().mockReturnValue({
        elementRef: 'elementRef',
        expanded: false,
        setExpanded: jest.fn(),
        triggerRef: jest.fn()
    })
}));

const defaultProps = {
    getStoreConfig: 'getStoreConfig'
};

const Component = props => {
    const talonProps = useStoreSwitcher(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...{ ...props, ...newProps }} />);

        return root.findByType('i').props.talonProps;
    };

    return { talonProps, tree, update };
};

test('should return correct shape', () => {
    const { talonProps } = getTalonProps(defaultProps);

    expect(talonProps).toMatchSnapshot();
});

describe('event handlers', () => {
    const { talonProps } = getTalonProps(defaultProps);

    test('handleSwitchStore switches store view', () => {
        const { handleSwitchStore } = talonProps;
        handleSwitchStore('store1');

        expect(setItem.mock.calls).toEqual([
            ['store_view_code', 'store1'],
            ['store_view_currency', 'USD']
        ]);
        expect(history.go).toHaveBeenCalledTimes(1);
    });

    test('handleSwitchStore does nothing when switching to not existing store', () => {
        const { handleSwitchStore } = talonProps;
        handleSwitchStore('store404');

        expect(setItem).toHaveBeenCalledTimes(0);
        expect(history.go).toHaveBeenCalledTimes(0);
    });
});
