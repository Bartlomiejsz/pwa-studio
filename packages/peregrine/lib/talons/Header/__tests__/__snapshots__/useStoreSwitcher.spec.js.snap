// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct shape 1`] = `
Object {
  "availableStores": Object {
    "store1": Object {
      "currency": "USD",
      "is_current": false,
      "locale": "locale1",
      "storeName": "Store 1",
    },
    "store2": Object {
      "currency": "EUR",
      "is_current": true,
      "locale": "locale2",
      "storeName": "Store 2",
    },
  },
  "handleSwitchStore": [Function],
  "handleTriggerClick": [Function],
  "storeMenuIsOpen": false,
  "storeMenuRef": "elementRef",
  "storeMenuTriggerRef": [MockFunction],
}
`;
